src/coreclr/vm/methodtablebuilder.cpp

```diff
@@ -8855,13 +8853,13 @@ MethodTableBuilder::HandleExplicitLayout(
 
 //*******************************************************************************
 // make sure that no object fields are overlapped incorrectly, returns the trust level
-/*static*/ ExplicitFieldTrust::TrustLevel MethodTableBuilder::CheckValueClassLayout(MethodTable * pMT, bmtFieldLayoutTag *pFieldLayout)
+/*static*/ ExplicitFieldTrust::TrustLevel MethodTableBuilder::CheckValueClassLayout(MethodTable * pMT, bmtFieldLayoutTag *pFieldLayout, UINT fieldBaseOffset)
```

Comment:
I do not think we need `fieldBaseOffset` arg. If there is a misaligned byref-like type inside the field type, loading of the field type is going to fail anyway, so we should not need to take the offset of the field into account when doing this check.

Reply:
> If there is a misaligned byref-like type inside the field type, loading of the field type is going to fail anyway

Perhaps I'm not understanding this statement, but adding support for a ByRefLike type to not be aligned on a pointer is what is happening in this PR. In this case, I think I need to know the offset of the field so it can be computed in the explicit offset case. I think the base offset is used in native AOT, how is CoreCLR different?

Reply:
What is the test case that is going to fail if you delete the `fieldBaseOffset` argument and replace its uses with `0`?


If I am understanding this correctly, the `fieldBaseOffset` argument is trying to handle this situation:
```
[SequentialLayout(Pack=1)]
ref struct A
{
    [FieldOffset(3)]
    B b;
}

[SequentialLayout(Pack=1)]
ref struct B
{
    [FieldOffset(5)]
    ref byte p;
}
```
The type A is technically valid since `ref byte p` inside type A is aligned and the `fieldBaseOffset` argument is trying to make it work. My point is that we should have failed to load the type `B` earlier since it has mis-aligned byref, so we do not need to make the effort to make the type A technically valid.

Reply:
> What is the test case that is going to fail if you delete the fieldBaseOffset argument and replace its uses with 0?

I need to propagate the field offset to the inner so `p` is computed to be at an offset of `4`. This would be fine on x86, but fail on x64. It is one of the tests I added.

```csharp
[SequentialLayout(Pack=1)]
ref struct A
{
    [FieldOffset(4)]
    B b;
}

ref struct B
{
    ref byte p;
}
```

Reply:
Got it. I see why this arg is needed now.